api.go

```diff
@@ -725,7 +725,9 @@ func (c Client) newRequest(method string, metadata requestMetadata) (req *http.R
 			// Gather location only if bucketName is present.
 			location, err = c.getBucketLocation(metadata.bucketName)
 			if err != nil {
-				return nil, err
+				if err != io.EOF {
```

Comment:
This is not a correct fix @hanqingwu we need to see why Aliyun is not returning a body. This means that Aliyun is not S3 compatible - can you ask Aliyun to fix this?

Reply:
this is the package  i capture .   there is no location  data in the  return message 

GET /testbucket/?location= HTTP/1.1
Host: cooper-wu-personal.oss-cn-shenzhen.aliyuncs.com
User-Agent: MinIO (linux; amd64) minio-go/v6.0.50
Authorization: AWS4-HMAC-SHA256 Credential=LTAI4FjoD1doA6SBRnsayY7z/20200320/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=89d1c26f9545df22c4c08994740fda335242841626cd7384800fbbf114bf6633
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20200320T055627Z

HTTP/1.1 200 OK
Server: AliyunOSS
Date: Fri, 20 Mar 2020 05:56:27 GMT
Content-Type: application/octet-stream
Content-Length: 0
Connection: keep-alive
x-amz-request-id: 5E745B0B1A6A0E33345E14C1
Accept-Ranges: bytes
ETag: "D41D8CD98F00B204E9800998ECF8427E"
Last-Modified: Thu, 19 Mar 2020 10:12:18 GMT
x-oss-object-type: Normal
x-oss-hash-crc64ecma: 0
x-amz-storage-class: STANDARD
Content-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==

Reply:
Yes this is wrong, if region is default the XML should be still sent..

Reply:
Hi harshavardhana ,  I have checked , Aliyun has supported getbucketlocation

GET /?location HTTP/1.1
Host: cooper-wu-personal.oss-cn-shenzhen.aliyuncs.com
User-Agent: aliyun-sdk-go/1.9.2 (Linux/4.15.0-54-generic/x86_64;go1.14)
Authorization: OSS LTAI4FjoD1doA6SBRnsayY7z:h86M7GU4k2UYIX81Cbt3/mq/cBw=
Date: Tue, 24 Mar 2020 06:04:55 GMT
Accept-Encoding: gzip

HTTP/1.1 200 OK
Server: AliyunOSS
Date: Tue, 24 Mar 2020 06:04:56 GMT
Content-Type: application/xml
Content-Length: 96
Connection: keep-alive
x-oss-request-id: 5E79A3083D174B3630BD23A9
x-oss-server-time: 60

<?xml version="1.0" encoding="UTF-8"?>
<LocationConstraint>oss-cn-shenzhen</LocationConstraint> 


The problem is Aliyun only supported virtual hosted style path.

Reply:
GET /cooper-wu-personal/?location= HTTP/1.1
Host: oss-cn-shenzhen.aliyuncs.com
User-Agent: MinIO (linux; amd64) minio-go/v6.0.50
Authorization: AWS4-HMAC-SHA256 Credential=LTAI4FjoD1doA6SBRnsayY7z/20200324/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=3dd765a070832ed8432378f86358f1ae79ddb5c3b4d59d53f52e739522fadad9
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20200324T061957Z

HTTP/1.1 403 Forbidden
Server: AliyunOSS
Date: Tue, 24 Mar 2020 06:19:57 GMT
Content-Type: application/xml
Content-Length: 260
Connection: keep-alive
x-amz-request-id: 5E79A68DAA5E733032F2809E

<?xml version="1.0" encoding="UTF-8"?>
<Error>
<Code>SecondLevelDomainForbidden</Code>
<Message>Please use virtual hosted style to access.</Message>
<RequestId>5E79A68DAA5E733032F2809E</RequestId>
<HostId>oss-cn-shenzhen.aliyuncs.com</HostId>
</Error>